enhancements/declarative-index-config.md

```diff
@@ -0,0 +1,461 @@
+
+A new sub-command `update` will be introduced under `opm index`, which will take a folder containing json representations of packages as it's input along with an existing index image tag, and will push any update made to the representations up to the index container image.
+
+>Note: Pushing an update to one or more package representation would create a discrepancy between the data in the existing registry database, and what's represented in the json files. To ensure this discrepancy does not exists, pushing an update to one or more packages would also have to be accompanied by removing and then re-adding the package/s changed in the database during that update.   
+
+```bash
+$ opm index inspect --index=quay.io/some-namespace/my-index 
+$ cd my-index
+$ // edit a package inside the index
+$ cd ..
+$ opm index update my-index --tag=docker.io/some-namespace/my-index:latest
+```
+
+The `update` command will verify that the timestamp in the file inside `my-index` folder matches with the timestamp inside the current index image before pushing overriding the content of the container image with the latest changes.   
+
+> Note: For a seamless UX, the following things should be consider during implementation when the `opm index update` command is summoned: 
+> 1. Validate the files inside the folder given as input to the command to ensure they are valid json representation of indexes (and not random files with random content, which could be potential security risks)
+> 2. If the files are valid files, verify that the timestamp locally matches the timestamp in the container image from the registry it's hosted on. If they do not match, throw an error. 
+> 3. If the files are valid, and the timestamp is valid, take a git diff between the present content of the index (i.e the existing json representations) and the content of the folder given as input. Display the diff as an output of the command with the following message 
+> "The following changes will be updated in the index:"
+> If there is no diff between the content in the container image and the content of the local folder, throw an error with the message "No diff found between index and local content to update."
+
+##### Introducing a new `validate` sub-command for `opm index`
+
+A new sub-command `validate` will be introduced under `opm index`, that will validate the content of each package's json representation. These validations will be in the context of rules about the structure of a package in an index (required fields, naming conventions etc). A best effort is currently being made to aggregate these rules from various sources and maintained in the [api project's validation library](https://github.com/operator-framework/api/tree/master/pkg/validation). The `opm index validate` will be responsible for validating the index for authors to make sure that the packages in the index conforms to the rules mentioned above. 
+
+#### Package config CUE definition
```

Comment:
I still question the requirement to use `cuelang` here. It seems like it might be a useful tool, but can't we start with something much simpler? A json file or set of files that describes the interface? Building this with cue seems like it increases the scope of this problem quite a bit without necessarily offering something that most of our users will need.

Reply:
@kevinrizza I'm trying to understand how using `cuelang` could increase the scope. From what I can see, using `cue` will actually help us leverage a lot of the functionality that it provides out of the box (defining the structure of a bundle/package/channel etc, validating the json files), that we'd have to implement and maintain otherwise. I guess to me using `cuelang` or a similar tool IS the simplest way to go about this.

Reply:
I understand where you're coming from. I definitely think we should have the plain JSON or YAML examples first. That being said, my read is that this is a suggested approach that keeps us out of supplying first-class validation features, which is a net gain for us maintainers.

Reply:
Circling back on this. Adding explicit support for `cue` in this enhancement ties this project to another project that potentially puts us on the hook for future support if that project ever loses its maintainers.

Furthermore, relying in a first class way on that project definitely increases scope. I'm still not convinced that having two layers of translation is a valuable abstraction. Are we really saying that there is a layer of abstraction that describes a package that our users just can't grok by reading raw json or yaml?

I see cue as an interesting abstraction that could help eventually with creating a flexible abstraction for our users to define arbitrary queries about subsets of data in the index if we ever wanted to expose a feature like that. I *don't* see a great argument for using cuelang to define our internal schema as part of an internal translation, similar to what we get now with the database structure of the sqlite database. We can get more or less the same thing with some imperative code that validates against a JSON object after we've constructed it. I don't really buy the argument that our schema is complex enough that we have that level of concern that only a database engine or a syntax engine can handle it for us. The moment we hit a bug with cuelang IMO all of this "value add" sort of goes out the window, and then we're still potentially on the hook for maintaining a small relatively untested project.

If the goal of this enhancement is just to build something that allows our users to read their package from the index as a blob so they can parse it, get interesting useful information from it, and pass it back in after making small modifications, why not just use raw JSON? That seems to simplify the problem significantly.

Reply:
@kevinrizza I see your point. The purpose of discussing CUE in this enhancement wasn't to propose CUE as the defacto syntax engine that lets us handle the final raw JSON. It is discussed as a potential project that can give us a head start by not having to build all the schema definition and validation logic ourselves. I get how the previous heading "Using CUE to create the package representation" is misleading, and have updated the enhancement with the title "Creation of the package json/yaml representation" where I discuss similar projects and an example to help us get started [here](https://github.com/operator-framework/enhancements/pull/37/files#diff-1edbfba9fdddd2b74b5cd32c6de146a4f62e1400a93bee78aed2cbf6ebaaa1d7R422-R427)